### PR TITLE
[teleport] Initial release of Teleport integration

### LINKED ISSUE #9478: [New Integration] Teleport Access Audit Events
### Description

"Teleport provides connectivity, authentication, access controls and audit for infrastructure.

It includes an identity-aware access proxy, a CA that issues short-lived certificates, a unified access control system and a tunneling system to access resources behind the firewall.

Teleport understands the SSH, HTTPS, RDP, Kubernetes API, MySQL, MongoDB and PostgreSQL wire protocols, plus many others. It can integrate with Single Sign-On providers and enables you to apply access policies using infrastructure-as-code and GitOps tools." ([source](https://goteleport.com/docs/))

"Teleport's Event Handler plugin receives audit events from the Teleport Auth Service and forwards them to your log management solution, letting you perform historical analysis, detect unusual behavior, and form a better understanding of how users interact with your Teleport cluster." ([source](https://goteleport.com/docs/management/export-audit-events/))

### Architecture
Teleport provides an Event Handler [plugin](https://goteleport.com/docs/management/export-audit-events/) to export audit events.

# Integration release checklist

This checklist is intended for integrations maintainers to ensure consistency
when creating or updating a Package, Module or Dataset for an Integration.

### All changes

- [x] Change follows the [contributing guidelines](https://github.com/elastic/integrations/blob/master/CONTRIBUTING.md).
- [ ] Supported versions of the monitoring target are documented.
- [x] Supported operating systems are documented (if applicable).
- [x] Integration or [System tests](https://github.com/elastic/elastic-package/blob/master/docs/howto/system_testing.md) exist.
- [x] Documentation exists.
- [x] Fields follow [ECS](https://github.com/elastic/ecs) and [naming conventions](https://www.elastic.co/guide/en/beats/devguide/master/event-conventions.html).
- [ ] At least a manual test with ES / Kibana / Agent has been performed.
- [x] Required Kibana version set to: 8.13.0+

### New Package

- [ ] Screenshot of the "Add Integration" page on Fleet added.

### Dashboards changes

- [x] Dashboards exists.
- [x] Screenshots added or updated.
- [x] Datastream filters added to visualizations.

### Log dataset changes

- [x] [Pipeline tests](https://github.com/elastic/elastic-package/blob/master/docs/howto/pipeline_testing.md) exist (if applicable).
- [x] Generated output for at least 1 log file exists.
- [x] Sample event (`sample_event.json`) exists.

### PR DESCRIPTION
**Pull request by ilyannn**
# Proposed commit message:

```
Implements the Teleport integration. 

  - Events are categorized with event.category, event.type and event.outcome.
  - Fields are renamed as appropriate:
    - Elastic Common Schema fields are used if possible; process and http groups are extended.
    - Most fields are classified into 29 custom field subgroups under "teleport.audit" group. 
    - Additional fields - "input.type" and log group - are used for the agent metadata.
    - Descriptions for custom fields are taken from the Teleport source code.
    - Logic depends of the event type, so "path" goes to "file.path", "process.executable" or "url.path".
    - Related fields are filled out for usernames, IP addresses and hosts.
  - Tests are included for each of the known event codes.
    - They are based on Teleport project fixtures.
    - IP addresses in the logs are changed in accordance with Elastic whitelist.
    - Fixes are included in some cases where fixtures were inconsistent with the code.
  - Fields of the form "198.51.100.196:50332" are split into "*.ip"/"*.domain" and "*.port".
    - Geo-information is added to these fields.
  - Enumerations are humanized, for example "teleport.audit.sftp.action" = "CLOSE" instead of 2.
  - Includes a sample dashboard that displays events with different slicing and a map.

See https://github.com/elastic/integrations/pull/10006 for sources and notes on implementation.
```

# Lists of tasks

### This is unclear:

- [x]  ~Should we include the ECS fields like `log.file.device_id` or `client.ip` in the `fields/` folder?~
- [x]  ~Should the failed login have two `event.types`: `start` and `error`, or just one `error`?~
- [x]  ~The `*.address` fields should be copied into either the `*.domain` or `*.ip` field according to the docs, is there an example processor somewhere?~
- [x]  ~Is there a place to put the information about building the integration (which might be useful for any updates), like `CONTRIBUTING.md`?~
- [x] ~pipeline test: `unexpected pipeline error: access to db denied. User does not have permissions. Confirm database user and name.` – this is not a pipeline error, it's an event with `error.message` filled, what to do?~
- [x] ~pipeline test: `field "teleport.audit.headers.Accept" is undefined` – yes, it's a map of headers, what's the correct field type?~

- [x] ~Some fields that clearly should be grouped with existing ECS fields aren't present in the schema, e.g.:~

- `process.flags`
- `process.cgroup_id`
- `http.request.headers`

~so I had to namespace them under `teleport.audit`. Is that a good solution or should I just use these fields directly, like in the [APM integration](https://github.com/elastic/integrations/pull/8881/files#diff-fc567aed173b96cec734583bad0697cc2f79657ccc3f86087cd7c0baa5a089f9R6)?~

- [x] ~Serialization of JSON here is wrong: is there a way (script) to serialize the JSON from some `field` into the string correctly? If not, how can I check if the field contains a string or an object (distinguish `'field': 'value'` from `'field': {'key': 'value'}`)?~

```json
            "http": {
                "request": {
                    "body": {
                        "content": "{TableName=test-table}"
                    },
```

### This is work to be fixed:

- [x]  Based on the event code it should be possible to fill out the failed outcome.
- [x]  Some fields, like `path`, should go into different groups (URL path or file path) depending on category.
- [x]  In `address` fields, the IPv6 addresses are included within `[` and `]`, which does not seem to work as ECS's `*.ip`.
- [x]  Sometimes `*.address` fields should be copied into `*.domain` instead of always `*.ip`.
- [x]  Errors should be copied into separate fields, not `error.message`.
- [x]  Use flattened types for headers.
- [x]  Add skipped `AccessListReviewMetadata` class
- [x]  Correctly fill out all the `related` fields.
- [x]  Check the geo fields.
- [x]  Remove the remains of the old generated pipeline. 
- [x]  Remove commented out fields from the `fields/*` files.
- [x]  Separately lint the YAML files.
- [x]  Combine custom fields into a single file.
- [x]  Check if we have any ECS fields that can be deleted from the list.
- [x]  Regenerate sample event
- [x]  Manually review the output from all events
- [x]  Separately lint Markdown files.
- [x]  Switch to format from  https://github.com/elastic/sei-llm-langserve/pull/290
- [x]  Check in with Teleport people for reviews and suggestions.

### This needs to be implemented:

- [x] A sample dashboard
- [x] Screenshots
- [x] Sample event
- [x] Teleport README
- [x] Commit message
- [x] How to test locally

### Ideas for future work:

- [ ]  Add any missing events from the list of events, like `billing.*`
- [ ]  Check in with Elastic teams that use Teleport for suggestions.
- [ ]  Add missing type for events like `sftp`.
- [ ]  Consistently use a single YAML string format for `description: >` fields
- [ ]  Rerun Integration Assistant with (a sample of) all events and check categorization.
- [ ]  Add system test assertions, e.g. "all types are filled out".
- [ ]  Find a way to use Teleport-branded color #512FCA in the design.

# How it was built

Manually built from the ground up, also merging changes from https://github.com/elastic/integrations/pull/9709 and https://github.com/elastic/integrations/pull/9937, with heavy use of GenAI tools.

### Sources

- [Teleport icon](https://goteleport.com/static/favicon.svg)
- Teleport source:
  - [Events message structure](https://github.com/gravitational/teleport/blob/master/api/proto/teleport/legacy/types/events/events.proto)
  - [Description of event types](https://github.com/gravitational/teleport/blob/master/web/packages/teleport/src/services/audit/makeEvent.ts)
  - [List of example events for each event code](https://github.com/gravitational/teleport/blob/master/web/packages/teleport/src/Audit/fixtures/index.ts)

IP replacement:

```shell
sed -e "s/1\.1.1./67.43.156.1/g;s/2\.2.2./67.43.156./g;s/198.51.100./175.16.199./g;s/172.10.1./67.43.156./g;s/100.104.52.89/81.2.69.192/g;s/190.58.129.4/89.160.20.112/g;s/192.000.0.000/89.160.20.128/g;s/50.34.48.113/81.2.69.193/g;s/54-162-177-255/175-16-199-255/g;s/\[::1\]/2a02:cf40::/g;s/198.51.100./1.128.0./g" -i bak data_stream/audit/_dev/test/pipeline/test-teleport-all-events.log
```

Note: event fixtures are not always internally consistent, example:

image

 
### How the ingest pipeline was generated

Using OpenAI through Azure AI Studio to go through the generated text list of events (see below), we generated the sub-pipeline to assign each event to a category (`data_stream/audit/elasticsearch/ingest_pipeline/event-categories.yml`).

Using OpenAI and the Go source of event message structure, we generated the field groups and the sub-pipeline to assign each field (`data_stream/audit/elasticsearch/ingest_pipeline/event-groups.yml`).

### How the text list of events was built

```just
#!/usr/bin/env just --justfile

# Run the whole conversion pipeline.
convert: download
    cat makeEvent.ts | rg "type:" -A1 | sed "s/--//g" | sed s/\',//g | sed "s/    ....: '//g" | sed 's/"//g' | just convert-script | sort > events.txt

# Script to convert a list of pairs of strings into a an 'A:B' format.
convert-script:
    #!/usr/bin/env python3
    import fileinput

    key = value = None

    for line in fileinput.input():
        line = line.strip()
        if not line:
            assert not key and not value
        elif key:
            value = line
            print(f"{key}: {value}")
            key = value = None
        else:
            key = line

# Download the original file.
download:
    curl -o makeEvent.ts https://raw.githubusercontent.com/gravitational/teleport/0aa5285477d422e98bb72ccf42a4381e6fdce527/web/packages/teleport/src/services/audit/makeEvent.ts
```

### Other notes on building the pipeline

- Needed  `'^(%{IPORHOST:server.address}|\[%{IP:server.ip}\])(:%{POSINT:server.port:long})?$'`  to cover the IPv6 addresses displayed together with the port.

Guesses had to be made about the meaning of some fields such as `os_login`. For example, here it is rendered as part of process group:

```json
            "process": {
                "name": "bash",
                "pid": 2653,
                "user": {
                    "name": "root"
                }
            },
```

Some fields should logically have come into the ECS fields, but seemed to not work with tests, e.g. `error.message` or `labels`. This for example failed with `field "labels.teleport.dev/origin" is undefined` etc.

```json
            "labels": {
                "env": "local",
                "teleport.dev/origin": "dynamic"
            },
            "orchestrator": {
                "cluster": {
                    "name": "root"
                }
            },
```

# Standard template

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## How to test this PR locally

There is nothing special – run pipeline tests or run system tests and check the data in Kibana with 

```sh
elastic-package test system --defer-cleanup 2h -v
```

## Related issues

- Closes #9478.

### CODE REVIEW

**Comment by andrewkroh** [comment]
> Should we include the ECS fields like `log.file.device_id` or `client.ip` in the `fields/` folder?

If you require Elastic stack 8.14 then you can omit the ECS fields because of the [new ecs@mappings](https://docs.google.com/document/d/1I5GKUB8AWUdEhqNjhHE2CXDeUMHk3_kcMaI3xKbQ16g/edit) component template that is being added to all integrations.

This only applies to the fields that are in ECS (log.file.device_id is not IIRC) and for fields that are using the standard definition (e.g. if you have `event.module` which is in ECS, but you have changed the type to be `constant_keyword` then you must keep that declaration).

File: packages/teleport/data_stream/audit/elasticsearch/ingest_pipeline/event-categories.yml [hunk 1]
```diff
@@ -0,0 +1,611 @@
+---
+description: Pipeline for setting the categorization fields
+processors:
+  - set:
+      field: event.kind
+      value: event
+  # Events related to access list activity.
+  - set:
```
Comment (bhapas) [inline packages/teleport/data_stream/audit/elasticsearch/ingest_pipeline/event-categories.yml:8]: I think all these should be `append` processors with `allow_duplicates: flase` if the event holds multiple types / categories

File: packages/teleport/data_stream/audit/elasticsearch/ingest_pipeline/event-enrich.yml [hunk 2]
```diff
@@ -0,0 +1,273 @@
+---
+description: >
+  Pipeline to enrich Teleport audit fields with additional information.
+
+processors:
+  # Enrich address fields.
+  - grok:
+      field: client.address
+      ignore_missing: true
+      ignore_failure: true
```
Comment (bhapas) [inline packages/teleport/data_stream/audit/elasticsearch/ingest_pipeline/event-enrich.yml:10]: May be failure should not be ignored here?

File: packages/teleport/data_stream/audit/elasticsearch/ingest_pipeline/event-enrich.yml [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+      ignore_missing: true
+  - rename:
+      field: client.as.organization_name
+      tag: rename_client_as_organization_name
+      target_field: client.as.organization.name
+      ignore_missing: true
+  - grok:
+      field: server.address
+      ignore_missing: true
+      ignore_failure: true
```
Comment (bhapas) [inline packages/teleport/data_stream/audit/elasticsearch/ingest_pipeline/event-enrich.yml:42]: Same here.. May be failure should not be ignored

**Review by bhapas — COMMENTED** [review]
Overall looks good.

Not sure if we should ignore the failures from the grok processor. I think they still need to be caught? @efd6 WDYT?

File: packages/teleport/manifest.yml [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  license: "Elastic-2.0"
+description: "Collect logs from Teleport with Elastic Agent."
+type: integration
+categories:
+  - monitoring
+  - network
+  - security
+conditions:
+  kibana:
+    version: "^8.13.0"
```
[the remark this answers is not in the document]
Comment (bhapas) [reply packages/teleport/manifest.yml]: Old comment.. Obsolete

File: packages/teleport/_dev/build/docs/README.md [hunk 5]
```diff
@@ -0,0 +1,28 @@
+# Teleport Audit Events Integration
+
+[Teleport](https://goteleport.com/docs/) provides connectivity, authentication, access controls and audit for infrastructure.
+
+This integration ingests audit events from Teleport. You can use it to perform historical analysis, 
+detect unusual behavior, and form a better understanding of how users interact with your Teleport cluster.
```
[the remark this answers is not in the document]
Comment (ilyannn) [reply packages/teleport/_dev/build/docs/README.md:6]: It's the Oxford comma, which is generally recommended for clarity. Our style guide also [recommends using it](https://brand.elastic.co/302f66895/p/194a3b-writing-style-guide/t/60c357).

File: packages/teleport/data_stream/audit/elasticsearch/ingest_pipeline/event-categories.yml (hunk 1 above)
Comment (ilyannn) [reply packages/teleport/data_stream/audit/elasticsearch/ingest_pipeline/event-categories.yml:22]: I try to keep the logic simple – for each event there should be exactly one `set:` with `field: event.category` and one `field: event.type` in the main part of the ingest pipeline. This way you don't need to worry that some other processor will redefine or append to these fields. 

Would using `append` be  more appropriate?

**Review by ilyannn — COMMENTED** [review]
Thanks for the comments!

File: packages/teleport/_dev/build/docs/README.md [hunk 6]
```diff
@@ -0,0 +1,28 @@
+# Teleport Audit Events Integration
+
+[Teleport](https://goteleport.com/docs/) provides connectivity, authentication, access controls and audit for infrastructure.
```
Comment (andrewkroh) [inline packages/teleport/_dev/build/docs/README.md:3]:
```suggestion
[Teleport](https://goteleport.com/docs/) provides connectivity, authentication, access controls, and audit for infrastructure.
```

nit: personally would love to see an Oxford comma here (similar in style to the next paragraph)

**Review by andrewkroh — APPROVED** [review]
Nice work. It looks thorough.

File: packages/teleport/_dev/build/docs/README.md (hunk 6 above)
Comment (ilyannn) [reply packages/teleport/_dev/build/docs/README.md:3]: Makes sense! (Honestly I stole this from a description by another person...)

**Pull request merged**